routing/index_graph.hpp

```diff
@@ -98,27 +104,112 @@ class IndexGraph final
     return GetGeometry().GetRoad(segment.GetFeatureId()).GetPoint(segment.GetPointId(front));
   }
 
-  RouteWeight CalcSegmentWeight(Segment const & segment);
+  template <typename Parent>
+  bool IsRestricted(Parent const & parent,
+                    uint32_t parentFeatureId,
+                    uint32_t currentFeatureId, bool isOutgoing,
+                    std::map<Parent, Parent> & parents) const;
 
+  bool IsUTurnAndRestricted(Segment const & parent, Segment const & child, bool isOutgoing) const;
 private:
+
+  RouteWeight CalcSegmentWeight(Segment const & segment);
+
   void GetNeighboringEdges(Segment const & from, RoadPoint const & rp, bool isOutgoing,
-                           std::vector<SegmentEdge> & edges);
+                           std::vector<SegmentEdge> & edges, std::map<Segment, Segment> & parents);
   void GetNeighboringEdge(Segment const & from, Segment const & to, bool isOutgoing,
-                          std::vector<SegmentEdge> & edges);
+                          std::vector<SegmentEdge> & edges, std::map<Segment, Segment> & parents);
   RouteWeight GetPenalties(Segment const & u, Segment const & v);
 
   void GetSegmentCandidateForJoint(Segment const & parent, bool isOutgoing, std::vector<Segment> & children);
-  void ReconstructJointSegment(Segment const & parent, std::vector<Segment> const & firstChildren,
+  void ReconstructJointSegment(JointSegment const & parentJoint,
+                               Segment const & parent,
+                               std::vector<Segment> const & firstChildren,
                                std::vector<uint32_t> const & lastPointIds,
-                               bool isOutgoing, std::vector<JointEdge> & jointEdges,
-                               std::vector<RouteWeight> & parentWeights);
+                               bool isOutgoing,
+                               std::vector<JointEdge> & jointEdges,
+                               std::vector<RouteWeight> & parentWeights,
+                               std::map<JointSegment, JointSegment> & parents);
 
   std::shared_ptr<Geometry> m_geometry;
   std::shared_ptr<EdgeEstimator> m_estimator;
   RoadIndex m_roadIndex;
   JointIndex m_jointIndex;
-  RestrictionVec m_restrictions;
+  Restrictions m_restrictionsForward;
+  Restrictions m_restrictionsBackward;
   RoadAccess m_roadAccess;
   RoutingOptions m_avoidRoutingOptions;
 };
+
+/// \note |parent| can be fake.
```

Comment:
Давай комментарий перенесем о объявлению.

Reply:
нет, это в реализации, а не объявлении

Reply:
Если важно в реализации, то обычно пишем в теле метода.

Т.е. ты считаешь, что при использовании метода знать, что |parent| может быть fake, не важно? Кроме того, с ходу не понятно, что мы проверяем, можем ли мы пройти от |parentFeatureId| к |currentFeatureId|.

Reply:
```cpp
/// \brief Check, that we can go to |currentFeatureId|.
  /// We pass |parentFeatureId| and don't use |parent.GetFeatureId()| because
  /// |parent| can be fake sometimes but |parentFeatureId| is almost non-fake.
  template <typename Parent>
  bool IsRestricted(Parent const & parent,
                    uint32_t parentFeatureId,
                    uint32_t currentFeatureId, bool isOutgoing,
                    std::map<Parent, Parent> & parents) const;
```